### DISCUSSION TITLE
How to fix locations of objects in between environment resets?

### DISCUSSION THREAD

**Discussion by ameesh-shah** [PKU-Alignment/safety-gymnasium#87] [Q&A]
Hi,

In order to further simplify the tasks provided in Safety Gymnasium, I'd like to fix the locations of the objects that are sampled upon each reset. I looked into the source code, but it seems that there is no easy way to do this without either building a new placements dict (which are generated in full during running) or changing the Underlying object source code to not repeatedly sample a new layout upon each reset. Is there a way to fix the locations of objects so that they do not change after resetting the environment? Thanks in advance!

**Comment by muchvo** [comment]
Please refer to the [provided code](https://github1s.com/PKU-Alignment/safety-gymnasium/blob/HEAD/safety_gymnasium/bases/base_task.py). We've organized several boolean flags using dataclasses. You can find the relevant classes in the code, particularly this one:

```python
class MechanismConf:
    r"""Mechanism options.

    Starting position distribution.

    Attributes:
        randomize_layout (bool): If false, set the random seed before layout to constant.
        continue_goal (bool): If true, draw a new goal after achievement.
        terminate_resample_failure (bool): If true, end episode when resampling fails,
        otherwise, raise a python exception.
    """

    randomize_layout: bool = True
    continue_goal: bool = True
    terminate_resample_failure: bool = True
```

This configuration is an attribute `self.mechanism_conf` within the base class `BaseTask`. Therefore, you should add the line `self.mechanism_conf.randomize_layout = False` in the `__init__` method of the specific task class you're working on.

**Reply by ameesh-shah** [reply]
Thanks very much for your help!

**Reply by ameesh-shah** [reply]
Hi, sorry, one more question - I'd like to enforce certain boundaries for the agent (so that it can't go beyond a certain point in the map.) I noticed that the wall objects are specific to certain tasks that I'm not using, and that the .pos attribute of the agent is not mutable. How would I be able to enforce these boundaries on the map that the agent cannot go past?

**Reply by muchvo** [reply]
If you're looking to add walls to an environment, please refer to the implementation in [this class](https://github1s.com/PKU-Alignment/safety-gymnasium/blob/HEAD/safety_gymnasium/assets/geoms/sigwalls.py). Note that the walls implemented by this class are penetrable by default. To make them solid, you should remove the following lines:

```python
'contype': 0,
'conaffinity': 0,
```

By doing so, the walls will have a physical presence. You'll just need to reimplement a class according to your needs and integrate it into your specific task like so:

```python
self._add_geoms(Hazards(num=8, keepout=0.18))
```

Since walls are immovable objects, they fall under the 'geom' category. For more information on the categorization of objects, please refer to the documentation [here](https://www.safety-gymnasium.com/en/latest/components_of_environments/objects.html).

If you wish to directly change the position of an object, you would need to use MuJoCo's interface. You can find several functions that accomplish this [here](https://github1s.com/PKU-Alignment/safety-gymnasium/blob/HEAD/safety_gymnasium/world.py#L464-L477):

```python
    def body_com(self, name):
        # Get the center of mass of a named body in the simulator world reference frame.
        return self.data.body(name).subtree_com.copy()

    def body_pos(self, name):
        # Get the position of a named body in the simulator world reference frame.
        return self.data.body(name).xpos.copy()

    def body_mat(self, name):
        # Get the rotation matrix of a named body in the simulator world reference frame.
        return self.data.body(name).xmat.copy().reshape(3, -1)

    def body_vel(self, name):
        # Get the velocity of a named body in the simulator world reference frame.
        return get_body_xvelp(self.model, self.data, name).copy()
...
    def set_state(self, value):
        """
        Sets the state from a dictionary.

        Args:
            value (dict): The desired state.
            call_forward: Optionally call sim.forward(). This is called by default if
                the udd_callback is set.
        """
        self.data.time = value['time']
        self.data.qpos[:] = np.copy(value['qpos'])
        self.data.qvel[:] = np.copy(value['qvel'])
        if self.model.na != 0:
            self.data.act[:] = np.copy(value['act'])
```

Feel free to reach out if you have any further questions or require additional assistance.
